File: ios/WalletStore.swift
import Foundation

class WalletSession {
  let wallet: Wallet
  var online: Online?
  
  init(wallet: Wallet, online: Online? = nil) {
    self.wallet = wallet
    self.online = online
  }
}

class WalletStore {
  static let shared = WalletStore()
  
  private var sessions: [Int: WalletSession] = [:]
  private var nextId: Int = 1
  private let queue = DispatchQueue(label: "com.rgbsdkrn.walletstore")
  
  private init() {}
  
  func create(wallet: Wallet) -> Int {
    return queue.sync {
      let id = nextId
      nextId += 1
      sessions[id] = WalletSession(wallet: wallet)
      print("WalletStore: Created wallet session with id: \(id)")
      return id
    }
  }
  
  func get(id: Int) -> WalletSession? {
    return queue.sync {
      return sessions[id]
    }
  }
  
  func setOnline(id: Int, online: Online) {
    queue.sync {
      sessions[id]?.online = online
      print("WalletStore: Set online for wallet session id: \(id)")
    }
  }
  
  func remove(id: Int) {
    queue.sync {
      if sessions.removeValue(forKey: id) != nil {
        print("WalletStore: Removed wallet session with id: \(id)")
      }
    }
  }
  
  func clear() {
    queue.sync {
      sessions.removeAll()
      nextId = 1
      print("WalletStore: Cleared all wallet sessions")
    }
  }
}

File: android/src/main/java/com/rgbsdkrn/WalletStore.kt
package com.rgbsdkrn

import android.util.Log
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.rgbtools.Online
import org.rgbtools.Wallet

data class WalletSession(
  val wallet: Wallet,
  var online: Online?
)

object WalletStore {
  private const val TAG = "WalletStore"
  private val mutex = Mutex()
  private val sessions = mutableMapOf<Int, WalletSession>()
  private var nextId = 1

  suspend fun create(wallet: Wallet): Int = mutex.withLock {
    val id = nextId++
    sessions[id] = WalletSession(wallet, null)
    Log.d(TAG, "Created wallet session with id: $id")
    return id
  }

  suspend fun get(id: Int): WalletSession? = mutex.withLock {
    return sessions[id]
  }

  suspend fun setOnline(id: Int, online: Online) = mutex.withLock {
    sessions[id]?.online = online
    Log.d(TAG, "Set online for wallet session id: $id")
  }

  suspend fun remove(id: Int) = mutex.withLock {
    val session = sessions.remove(id)
    if (session != null) {
      Log.d(TAG, "Removed wallet session with id: $id")
    }
  }

  suspend fun clear() = mutex.withLock {
    sessions.clear()
    nextId = 1
    Log.d(TAG, "Cleared all wallet sessions")
  }
}

